File: app/src/main/java/com/irateam/vkplayer/api/SimpleCallback.java
package com.irateam.vkplayer.api;

public class SimpleCallback<T> implements Callback<T> {

    private SuccessListener<T> successListener;
    private ErrorListener errorListener;
    private FinishListener finishListener;

    protected SimpleCallback() {

    }

    public static <C> SimpleCallback<C> success(SuccessListener<C> listener) {
        SimpleCallback<C> callback = new SimpleCallback<>();
        callback.successListener = listener;
        return callback;
    }

    public SimpleCallback<T> error(ErrorListener errorListener) {
        this.errorListener = errorListener;
        return this;
    }

    public SimpleCallback<T> finish(FinishListener finishListener) {
        this.finishListener = finishListener;
        return this;
    }

    @Override
    public void onComplete(T result) {
        notifySuccess(result);
        notifyFinish();
    }

    @Override
    public void onError() {
        notifyError();
        notifyFinish();
    }

    private void notifySuccess(T result) {
        if (successListener != null) {
            successListener.onSuccess(result);
        }
    }

    private void notifyError() {
        if (errorListener != null) {
            errorListener.onError();
        }
    }

    private void notifyFinish() {
        if (finishListener != null) {
            finishListener.onFinish();
        }
    }

    public interface SuccessListener<T> {

        void onSuccess(T result);
    }

    public interface ErrorListener {

        void onError();
    }

    public interface FinishListener {

        void onFinish();
    }
}

File: app/src/main/java/com/irateam/vkplayer/api/SimpleCallback.kt
package com.irateam.vkplayer.api

import java.util.*

open class SimpleCallback<T> : Callback<T> {

	private val successListeners: ArrayList<((T) -> Unit)> = ArrayList();
	private val errorListeners: ArrayList<(() -> Unit)> = ArrayList();
	private val finishListeners: ArrayList<(() -> Unit)> = ArrayList();

	constructor()

	constructor(block: SimpleCallback<T>.() -> Unit) {
		block.invoke(this)
	}

	open fun onSuccess(successListener: ((T) -> Unit)): SimpleCallback<T> {
		successListeners.add(successListener)
		return this
	}

	open fun onError(errorListener: () -> Unit): SimpleCallback<T> {
		errorListeners.add(errorListener)
		return this
	}

	open fun onFinish(finishListener: () -> Unit): SimpleCallback<T> {
		finishListeners.add(finishListener)
		return this
	}

	/**
	 * Do not use this methods in constructor block.
	 * It is implementation of Callback interface.
	 */
	override fun onComplete(result: T) {
		notifySuccess(result)
		notifyFinish()
	}

	override fun onError() {
		notifyError()
		notifyFinish()
	}

	private fun notifySuccess(result: T) {
		successListeners.forEach { it.invoke(result) }
	}

	private fun notifyError() {
		errorListeners.forEach { it.invoke() }
	}

	private fun notifyFinish() {
		finishListeners.forEach { it.invoke() }
	}

}